xcode/Subconscious/Shared/Components/Common/MarkupTextViewRepresentable.swift

```diff
@@ -243,6 +230,39 @@ struct MarkupTextViewRepresentable: UIViewRepresentable {
                 )
             )
         }
+        
+        // MARK: - NSTextLayoutManagerDelegate
+                                
+        func textLayoutManager(_ textLayoutManager: NSTextLayoutManager,
+                               textLayoutFragmentFor location: NSTextLocation,
+                               in textElement: NSTextElement) -> NSTextLayoutFragment {
+            
+            // TODO: might be better to hold a ref to textContentStorage somewhere
+            // textContentStorage is a concrete implementation of textContentManager
+            let textContentStorage = textLayoutManager.textContentManager as! NSTextContentStorage
+            let content = textContentStorage.attributedString(for: textElement)
+            
+            // Where we decide which layout/rendering implementation to use per TextLayoutFragment
+            if renderTranscludeBlocks && content?.string.contains("/slashlink") ?? false {
+                let layoutFragment = TranscludeBlockLayoutFragment(textElement: textElement, range: textElement.elementRange)
+                layoutFragment.text = content?.string
+                return layoutFragment
+                
+            } else {
+                return NSTextLayoutFragment(textElement: textElement, range: textElement.elementRange)
+            }
+        }
+        
+        // MARK: - NSTextContentStorageDelegate
+        
+        func textContentStorage(_ textContentStorage: NSTextContentStorage, textParagraphWith range: NSRange) -> NSTextParagraph? {
+            // In this method, we'll inject some attributes for display, without modifying the text storage directly.
+            let originalText = textContentStorage.textStorage!.attributedSubstring(from: range)
+            let textWithDisplayAttributes = NSMutableAttributedString(attributedString: originalText)
+            
+            self.representable.renderAttributesOf(textWithDisplayAttributes)
+            return NSTextParagraph(attributedString: textWithDisplayAttributes)
```

Comment:
This callback should be invoked only for `NSTextParagraph`s that have changed, massively speeding up rendering the attibutes 🎊 

(`NSTextParagraphs` roughly correlate to `NSTextLayoutFragments` 1:1, with some exceptions)

Reply:
Let's log here.

```
MarkupTextViewRepresentable.logger.debug(
    "textContentStorage: render markup attributes"
)
```

MarkupViewRepresentable contains a lot of tricky lifecycle logic, and I like to have logging in key areas such as rendering, so we can easily verify that e.g. rendering happens only once per keystroke.